### PR TITLE
Qstr autogen

### PR DESCRIPTION
**Pull request by janbraiins**
Hello,

I have put together a rather non-intrusive set of patches that make the life with qstrings much easier. The patch that actually strips the qstrdefs{,port}.h files makes adjustment in the build system of the platforms comes last. I believe all this needs some more polishing - at least in terms of adding/fixing the remaining platforms - I am not able to test windows build and esp8266 build.

I am open to any comments/improvements.

Jan

### CODE REVIEW

**Comment by danicampora** [comment]
@honzik666 wow, thanks! :-) I think this is a very nice feature for our build system!

**Comment by danicampora** [comment]
As a quick comment, I'd squash commits: https://github.com/braiins/micropython/commit/8fc151c5dddd5ea6edd7658ec042cf81861eca5c and https://github.com/braiins/micropython/commit/862525218d89089d5bbb327d956995424a3af512 into one. 
So first you add the new script and the you make all the changes to the makefiles and manual qstrdefs.

**Comment by honzik666** [comment]
Hi Dan,

yes, I was planning to squash them, however, I thought it would be initially easier to discuss as the second patch (https://github.com/braiins/micropython/commit/862525218d89089d5bbb327d956995424a3af512af512) is very polluted due to removing things from qstrdefs files. But sure, I will do that

**Comment by honzik666** [comment]
the last 2 commits are now squashed, windows build is now fixed, too I think - I had no way to check it except for running it through your CI

**Comment by honzik666** [comment]
Below are results without/with the autogen enabled. These reveal some minor incosistencies in each port's ifdef's within manually defined qstring sections.
# cc3200
  text	   data	    bss	    dec	    hex	filename
 183976	   1216	  76952	 262144	  40000	build-orig/WIPY/release/application.axf

LINK build/WIPY/release/application.axf
   text	   data	    bss	    dec	    hex	filename
 183576	   1216	  77352	 262144	  40000	build/WIPY/release/application.axf

# teensy
LINK build/micropython.elf
   text	   data	    bss	    dec	    hex	filename
 191388	   1236	  20388	 213012	  34014	build/micropython.elf
LINK build/micropython.elf
   text	   data	    bss	    dec	    hex	filename
 191052	   1236	  20388	 212676	  33ec4	build/micropython.elf

# unix
   text	   data	    bss	    dec	    hex	filename
 392430	   1512	   3104	 397046	  60ef6	micropython

   text	   data	    bss	    dec	    hex	filename
 392286	   1512	   3104	 396902	  60e66	micropython

# stmhal
   text	   data	    bss	    dec	    hex	filename
 277096	    344	  27656	 305096	  4a7c8	build-STM32F4DISC/firmware.elf

  text	   data	    bss	    dec	    hex	filename
 277068	    344	  27656	 305068	  4a7ac	build-STM32F4DISC/firmware.elf

**Comment by honzik666** [comment]
The windows build is still broken, anybody with MSVC and it's build tool experience have an idea how to fix this?

**Comment by stinos** [comment]
I'll try to get it fixed today

**Comment by dpgeorge** [comment]
As it is, this can't be merged: every time you edit a file the entire source must be recompiled.  That's because the master qstr file is regenerated if you touch a file, and then everything depends on the master qstr file.

**Comment by dpgeorge** [comment]
See py/makeversionhdr.py to see how you can fix it: that script will generate a temp version header, and only replace the existing if something changed.

**Comment by stinos** [comment]
Had a very quick look already for the mingw port (building from cygwin commandline): cygwin has no /dev/stdout and /dev/stdin, not sure what the best way is to fix that. I split the piping in two commands, using a temporary file to get it running. All files are preprocessed now, looks like a build might take twice the amount of time though, and uPy builds.

**Comment by dpgeorge** [comment]
> looks like a build might take twice the amount of time though

Yes, I see the same.  Processing a source file to extract qstrs seems to take about the same time as compiling it with gcc -Os.

**Comment by honzik666** [comment]
You are right, the initial build takes longer. However, it regenerates the qstrdefs using dependencies so for normal development withing C-sources it has no additional time impact. It will get rebuild if you touch the configuration.

**Comment by honzik666** [comment]
@dpgeorge I have extended the script to taken additional arguments and allow it to detect changes in the output similarly to mpversion extraction script. 

This patch https://github.com/braiins/micropython/commit/18bc8d6f6be4179409e419b165d2935690bd4e7d is currently separate as I would like to know the general opinion here. If we add the mpconfig* dependencies, the issue is that the generating script would spend a considerable amount of time checking/comparing the qstrings without generating anything. The deficiency of GNU make to use timestamp based dependencies is a bit frustrating here.
If we remove it, chances are that a change in mpconfigport.h would break the build since the relevant qstrdefs won't get regenerated.

@stinos script and rules have been made portable - they don't use any linux specific stuff like /dev/std{in/out}.

**Comment by stinos** [comment]
It's ok for mingw builds now. I made a quick and dirty version of the generation with msbuild but the content of qstrdefs.generated.h isn't the same as before yielding a faulty executable. For instance, second line I get is

```
QDEF(MP_QSTR__star_, (const byte*)"\xb1\xc2\x06" "_star_")
```

while that should be

```
QDEF(MP_QSTR__star_, (const byte*)"\x8f\xb5\x01" "*")
```

Any idea where it goes wrong? Also lines are in different order, does that matter?

 I'm just running all source files and qstrdefs.h through the preprocessor. The output of  source file preprocessing is passed to makeqstrdefs.py. The output of that, plus the preprocessed qstrdefs.h is passed to makeqstrdata.py like before.

**Comment by dpgeorge** [comment]
> Also lines are in different order, does that matter?

No, order does not really matter.

**Comment by honzik666** [comment]
Perfect, we are getting there. Some of the strings are special and need to be explicitely defined in qstrdefsport.h (e.g. for unix: https://github.com/braiins/micropython/blob/qstr-autogen/stmhal/qstrdefsport.h). However, this header doesn't exist for the windows build.

As of the 'star' qstring issue, the string is explicitely contained in: https://github.com/braiins/micropython/blob/qstr-autogen/py/qstrdefs.h#L37. One of the reasons why you are missing the correct version in your qstrdefs.generated.h, could be the order how you feed the output to https://github.com/braiins/micropython/blob/qstr-autogen/py/makeqstrdata.py - This final processing step removes duplicities and in fact considers the strings that appeared sooner as the right ones. Therefore, the point is to have the qstrdefs.collected.h processed/fed to makeqstrdata.py (see https://github.com/braiins/micropython/blob/qstr-autogen/py/py.mk#L209)

**Comment by stinos** [comment]
> However, this header doesn't exist for the windows build.

windows ports are basically part of unix port + a posix/uPy compatibility layer where needed + build definitions, so they use unix/qstrdefsport.h.

I indeed had the order wrong, fixed that so qstrdefs.h comes first and all is fine: https://github.com/stinos/micropython/commit/aa6cf4fab0442a44e4a812f19053c8e6f949e7e1. You should be able to cherry-pick/amend that commit. Note it is slow, brute-force and using temporary files and always generates everything as I don't immediately have time to figure out how to preprocess just the modified files, which should be possible though.

**Comment by honzik666** [comment]
@stinos, perfect, I have cherry-picked your commit as is for now. I will wait for @danicampora  or @dpgeorge before squashing the past 2 commits into 1.
Thanks, I have no easy way to test the windows build, hopefully this time the CI will pass for windows.

**Comment by danicampora** [comment]
Great, Looks good to me!

**Comment by stinos** [comment]
> I have no easy way to test the windows build

Actually you can temporarily access the Appveyor VM instance running the build: http://www.appveyor.com/docs/how-to/rdp-to-build-worker. Quite nifty.

File: py/mkrules.mk [hunk 1]
```diff
@@ -46,10 +46,30 @@ vpath %.c . $(TOP)
...
 $(BUILD)/%.pp: %.c
 	$(ECHO) "PreProcess $<"
 	$(Q)$(CC) $(CFLAGS) -E -Wp,-C,-dD,-dI -o $@ $<
 
+$(EMPTY_QSTRDEFS_GENERATED_H):
+	$(ECHO) "Generate empty $@ to satisfy qstring generator"
+	$(Q)mkdir -p $(dir $@)
+	$(Q)touch $@
+
 # The following rule uses | to create an order only prereuisite. Order only
```
Comment (hosaka) [inline py/mkrules.mk]: prerequisite*

**Comment by dpgeorge** [comment]
@honzik666 please leave @stinos 's commit unsquashed.

One problem with removing (most) qstrs from py/qstrdefs.h is that you now force all ports to auto-generate.  Some ports can't do this (eg microbit) and having a pre-generated py/qstrdefs.h file is important.

**Comment by honzik666** [comment]
@dpgeorge I will leave that as is. Just wondering what is microbit port? Did you mean pic16bit - I forgot to include it into the changes, this can be quickly fixed, though? I don't understand why a port wouldn't be able to perform this auto-generation step if windows port was able to do that with msbuild?

**Comment by dpgeorge** [comment]
microbit is a separate repository: https://github.com/bbcmicrobit/micropython .  They have a source-code dump of the relevant files from this repo, and us the "yotta" build system.  It's not easy to incorporate auto-qstr generation into yotta.  Even if it was, I'd still like to keep it so you can build without needing to run the auto-qstr script.

**Comment by honzik666** [comment]
@dpgeorge I see. Well, if we disable autogeneration we are back to where we were before - maintaining the definitions and introducing unintentional inconsistencies through not-up-to-date ifdefs. 
The autogeneration itself was not intrusive as long as you keep the full qstrdefs.h and qstrdefsport.h for a port that doesn't want to use. Therefore my suggestion is then to put py/qstrdefs.h where it was before stripping it. I am assuming microbit has it's own qstrdefsport.h and just uses the standard way.

Would this be acceptable? Actually - I had originally the 2 commits separate - one has introduced autogeneration and the second one has stripped the qstrdefs.h

**Comment by dpgeorge** [comment]
@honzik666 It might be ok in the end: everyone using uPy has to anyway run the makeqstrdata.py script to generate the actual data (and also makeversionhdr.py to generate the version header file), so they will need to have Python running, and then it should be ok to run makeqstrdefs.py.

I will need to take a look at this PR in detail and think about it more.  Thanks for your hard work on this!

**Comment by honzik666** [comment]
@dpgeorge any time, let me know if any further cleanup/rework is needed.

I have more things in the pipeline for uPy, this  resulted from our research here @braiins. Well, thanks also @prusnak for starting this initiative ;-)

**Comment by dhylands** [comment]
It might make sense to have a "no-autogen" for some targets. This would mean you run the autogen script manually and checkin the qstrdefs files (in a target specific location).

**Comment by honzik666** [comment]
Dave, I have added a control variable for such purpose. You can now explicitely disable qstr autogeneration or override the variable within the particular Makefile as needed.

**Comment by dpgeorge** [comment]
@honzik666 a few questions:

- It still rebuilds everything if I touch only 1 file (eg in stmhal: make; touch main.c; make); enabling debug output in makeqstrdefs it prints "File HAS changed, overwriting".  What is wrong?

- I see that you run each source file through the preprocessor before piping it into makeqstrdefs (what is the reason for this?  This is one of the main reasons why the extraction of qstrs is slow.).  That means there are no preprocessor directives left in the source file (no #if's for example).  So why does your makeqstrdefs script need to look for directives?

- What is the reason for the special handling of nlr* and emitn*?

**Comment by dhylands** [comment]
I haven't had a chance to look at this, but there command line options you can use to tell the preprocessor to preserve the directives.

**Comment by stinos** [comment]
Another thing: why do the qstr files still contain (in some cases, thousands of) preprocessor lines and not just Q() lines? It makes them unnecesarily large.

**Comment by prusnak** [comment]
Because you don't want to parse and evaluate #ifdefs in generator
written in Python. That's C preprocessor's job.
 
-- 
Best Regards / S pozdravom,
 
Pavol Rusnak <stick@gk2.sk>
 
 
On Thu, Mar 17, 2016, at 09:39 AM, stinos wrote:
> Another thing: why do the qstr files still contain (in some cases, thousands of) preprocessor lines and not just Q() lines? It makes them unnecesarily large.
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub[1]
>
 

Links:

  1. https://github.com/micropython/micropython/pull/1899#issuecomment-197763934

**Comment by dpgeorge** [comment]
> Because you don't want to parse and evaluate #ifdefs in generator written in Python. That's C preprocessor's job.

Then why does makeqstrdefs.py parse the #if's?  They don't exist after going through CPP.

**Comment by honzik666** [comment]
@dpgeorge well, let me explain this a bit - I have started the script so that it parses ifdefs and does all the hard work. However, it turns out that it is not possible to simple contatenate such output from all output files and feed it into the standard qstrdefs processing. The problem is that the concatenated output is not valid as a single file for CPP that is used as one of the steps in qstrdefs.generated.h.

So, I have reversed the order -> I have made full preprocessing of every file and then have used the script to extract the qstrdefs. The preprocessor token parsing logic has stayed in place.

As of the undesired rebuild of all files - I have justed tested it and it seems to behave exactly like we want. Just wondering, did you refresh the branch in your working copy? I have forced the base changeset few days ago. Well, to prevent any further confusion, I will be committing only linear fixes and if we arrive at some final variant, I will squash whatever changes you like.

`make  BOARD=STM32F4DISC 2>&1 | tee buildlog7.txt
Use make V=1 or set BUILD_VERBOSE in your environment to increase build verbosity.
Skip regeneration of: build-STM32F4DISC/genhdr/main.c.qstr
CC main.c
LINK build-STM32F4DISC/firmware.elf
   text	   data	    bss	    dec	    hex	filename
 277068	    344	  27656	 305068	  4a7ac	build-STM32F4DISC/firmware.elf
Create build-STM32F4DISC/firmware.dfu
Create build-STM32F4DISC/firmware.hex
```

**Comment by dpgeorge** [comment]
> The problem is that the concatenated output is not valid as a single file for CPP that is used as one of the steps in qstrdefs.generated.h.

What is the reason for this?

Actually, it's probably better that CPP does all the processing at the very first step, because there can be some tricky cases to parse #'s, eg if they are commented out.  Probably therefore more sustainable for makeqstrdefs.py to be as simple as possible, ie no # processing.

The reason that it rebuilds everything for me when I touch main.c is that I'm using Python 3, and the string comparison to check if a file changed compares str with bytes (and always returns False).  I guess you only checked it using Python 2?

Note that stmhal build *increases* by about 230 bytes with the auto-generation.  The reason is the ordering of qstrs has changed: without auto-gen, qstrs for the core are listed first (hence their id is small, <256), and small machine instructions can be used to encode them.

One other problem is that you cannot deduce the original string given the qstr name.  Eg MP_QSTR_abc_lt_ could have come from "abc_lt_" or "abc<".  So auto-gen only works for strings without escaped characters.

**Comment by honzik666** [comment]
This has been already taken care of, when you refresh your working copy
the preprocessor directives will be gone.
 It's my fault that I have reworked/forced the branch multiple times..


 Dne Thu, 17 Mar
2016 01:39:26 -0700 stinos <notifications@github.com> napsal(a):

> Another thing: why do the qstr files still contain (in some cases,
> thousands of) preprocessor lines and not just Q() lines? It makes
> them unnecesarily large.
> 
> ---
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/micropython/micropython/pull/1899#issuecomment-197763934

-- 
CTO
Braiins Systems
tel: +420 604 566 382
email: jan.capek@braiins.cz
web: http://www.braiins.cz

**Comment by honzik666** [comment]
Dne Thu, 17 Mar 2016 07:25:15 -0700
Damien George <notifications@github.com> napsal(a):

> > The problem is that the concatenated output is not valid as a
> > single file for CPP that is used as one of the steps in
> > qstrdefs.generated.h.
> 
> What is the reason for this?
> 
> Actually, it's probably better that CPP does all the processing at
> the very first step, because there can be some tricky cases to parse
> #'s, eg if they are commented out.  Probably therefore more
> sustainable for makeqstrdefs.py to be as simple as possible, ie no #
> processing.
Yes, exactly, that is the reason and has been experimentally verified
that correct processing is very tricky. However, I was able to get
there.
> 
> The reason that it rebuilds everything for me when I touch main.c is
> that I'm using Python 3, and the string comparison to check if a file
> changed compares str with bytes (and always returns False).  I guess
> you only checked it using Python 2?
Ooops, sorry, but that string compare is the only thing taken over from
the py/makeversionhdr.py script so I didn't assume there could be
anythin wrong with it, my bad, I will fix it.
> 
> Note that stmhal build *increases* by about 230 bytes with the
> auto-generation.  The reason is the ordering of qstrs has changed:
> without auto-gen, qstrs for the core are listed first (hence their id
> is small, <256), and small machine instructions can be used to encode
> them.
We could enforce some sort of ordering - e.g. core goes first if you
like too. I am a bit surprised by the 230 bytes you are mentioning. I
have verified the build with/without (using master branch at the branch
point where qstr-autogen has been started):
```
AUTOGEN DISABLED:
text	   data	    bss	    dec hex	filename
277096	    344	  27656 305096	  4a7c8

AUTOGEN ENABLED:
text	   data bss	    dec	    hex filename 
277068     344	  27656	 305068 4a7ac
```

Where exactly do you see the 230 byte increase? The text section is
smaller due to elimination of unused qstrings due to incosistencies
between the qstrdefsport.h and the code that uses the strings.
> 
> One other problem is that you cannot deduce the original string given
> the qstr name.  Eg MP_QSTR_abc_lt_ could have come from "abc_lt_" or
> "abc<".  So auto-gen only works for strings without escaped
> characters.
Yup, that's the reason why those special qstrings still need to be
explicitely listed - see the reduced qstrdefs.h and qstrdefsport.h. I
still find it easier just to define a couple special strings this way
than requiring to maintain huge headers.
> 
> ---
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/micropython/micropython/pull/1899#issuecomment-197901041

-- 
CTO
Braiins Systems
tel: +420 604 566 382
email: jan.capek@braiins.cz
web: http://www.braiins.cz

**Comment by prusnak** [comment]
Ping. Are there any other issues that need to be resolved?

**Comment by honzik666** [comment]
Dne Mon, 28 Mar 2016 13:05:47 -0700
Pavol Rusnak <notifications@github.com> napsal(a):

> Ping. Are there any other issues that need to be resolved?
Not that I know of. The latest issue with running the makeqstrdefs
script in python3 has been resolved.
> 
> ---
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/micropython/micropython/pull/1899#issuecomment-202561349

--

**Comment by pfalcon** [comment]
Sorry guys, we're very busy working on esp8266 port, so it may take extra time to get back to this.

**Comment by prusnak** [comment]
@honzik666 Big changes landed in master and they broke your PR. Could you please rebase?

**Comment by honzik666** [comment]
Done
Dne Thu, 14 Apr 2016 06:55:48 -0700
Pavol Rusnak <notifications@github.com> napsal(a):

> @honzik666 Big changes landed in master and they broke your PR. Could
> you please rebase?
> 
> ---
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/micropython/micropython/pull/1899#issuecomment-209954552

-- 
CTO
Braiins Systems
tel: +420 604 566 382
email: jan.capek@braiins.cz
web: http://www.braiins.cz

**Comment by dpgeorge** [comment]
Merged!  I rebased and squished some commits, made a few minor edits and cleaned up the commit messages.  See dbbf0827862d2e300827468020c6c558b7934276 through 73ccb3fc5b72d48d0b7370252f0d5126dedc5cba.

The biggest change I made was to remove the unused preprocessor-directive processing code from py/makeqstrdefs.py.  So that file is now very simple.

The only thing left that is not optimal is the following: if you edit a file but don't change any qstrs in it then makeqstrdefs.py will be run on that file each time you run "make", even if there is nothing to do.  So it can get to a stage after a while editing many files that a "make" will take some time to complete, even though it does "nothing".  This can be fixed later on if it becomes a real problem.

Great work everyone!

File: py/mkrules.mk (hunk 1 above)
Comment (dpgeorge) [inline py/mkrules.mk]: @hosaka this was not fixed because it's not part of this PR, it was there from long ago.

**Comment by pfalcon** [comment]
> As it is, this can't be merged: every time you edit a file the entire source must be recompiled. That's because the master qstr file is regenerated if you touch a file, and then everything depends on the master qstr file.

Yes, it's double-catch situation: currently, clean rebuild takes much longer, because files are preprocessed one by one. It would be much faster to pass all files to gcc -E and process everything en-masse, but then there would be larger delay with incremental builds.

**Comment by dpgeorge** [comment]
> It would be much faster to pass all files to gcc -E and process everything en-masse

Might be possible to pass just the files that have changed to gcc -E... but then it's harder to check what qstrs changed in the output.  But I don't know if this would help: probably the thing taking the most time is loading the file from disk, which must be done even if you pass all the files together.

In general solving the problem is hard because all files now depend on all files.  And we need to use CPP twice as much as before.

Another option is to keep the core qstrs predefined, by hand (as they were before) and just use auto qstr generation for the ports.  That way you cut out a large majority of source files that need processing.

**Comment by pfalcon** [comment]
I  see the same issue as reported before by @dpgeorge : change to a single file leads to all qstr files to be rebuilt:

~~~
$ time make
Use make V=1 or set BUILD_VERBOSE in your environment to increase build verbosity.
CC moduos.c
moduos.c:117:1: error: no return statement in function returning non-void [-Werror=return-type]
make: *** [build/moduos.o] Error 1
real	0m44.381s
~~~

Underneath:
~~~
$ time make V=1
python ../py/makeversionhdr.py build/genhdr/mpversion.h
xtensa-lx106-elf-gcc -E  -P -DN_X64 -DN_X86 -DN_THUMB -DN_ARM -Ibuild/tmp -I../lib/netutils -I. -I.. -I../stmhal -I../lib/mp-readline -I../lib/netutils -I../lib/timeutils -Ibuild -I/mnt/hdd/projects-3rdparty/xtensa/esp-open-sdk/xtensa-lx106-elf/bin/../xtensa-lx106-elf/sysroot/usr/include -Wall -Wpointer-arith -Werror -ansi -std=gnu99 -nostdlib -DUART_OS=0 -fsingle-precision-constant -Wdouble-promotion -D__ets__ -DICACHE_FLASH -fno-inline-functions -Wl,-EL -mlongcalls -mtext-section-literals -mforce-l32 -DLWIP_OPEN_SRC -DMICROPY_PY_USSL=1 -I../lib/axtls/ssl -I../lib/axtls/crypto -I../lib/axtls/config -Os -DNDEBUG  -fdata-sections -ffunction-sections ../py/mpstate.c -o - | python ../py/makeqstrdefs.py -s -o build/genhdr/py/mpstate.c.qstr
...
~~~


It was brave move to merge this 2 weeks before esp8266 release with a lot of work to do, and can we please have this reverted?

**Comment by pfalcon** [comment]
> Might be possible to pass just the files that have changed to gcc -E... but then it's harder to check what qstrs changed in the output.

Not harder: drop -P switch, split output based on filename info in preprocessor output, dump to temporary files, don't overwrite existing files if nothing changed (alternatively, reset timestamp).

> But I don't know if this would help: probably the thing taking the most time is loading the file from disk, which must be done even if you pass all the files together.

It's well known (?) that what takes long time is process startup/teardown, that's what to be avoided.

**Comment by dpgeorge** [comment]
> I see the same issue as reported before by @dpgeorge : change to a single file leads to all qstr files to be rebuilt:

A `make clean; make` should fix it.  Then if you edit a single file it only needs to preprocess that file.

**Comment by honzik666** [comment]
Thank you, I believe, this is a big step forward.

Cheers,

Jan

**Comment by dpgeorge** [comment]
> Thank you, I believe, this is a big step forward.

Actually, after using it for a few days I'm starting to experience more and more problems with it.  One of the biggest problems (as @pfalcon found) is that if you edit mpconfigport.h then every subsequent invocation of make takes ages.  Because the config changed, the qstrs might have changed, and so it needs to regenerate all .c.qstr files (that's expected).  But if none of the qstrs actually change then it doesn't update the .c.qstr file.  And then next time round you run make it still tries to generate all qstr files again.

**Comment by honzik666** [comment]
Mon, 18 Apr 2016 04:04:40 -0700
Damien George <notifications@github.com> wrote:

> > Thank you, I believe, this is a big step forward.
> 
> Actually, after using it for a few days I'm starting to experience
> more and more problems with it.  One of the biggest problems (as
> @pfalcon found) is that if you edit mpconfigport.h then every
> subsequent invocation of make takes ages.  Because the config
> changed, the qstrs might have changed, and so it needs to regenerate
> all .c.qstr files (that's expected).  But if none of the qstrs
> actually change then it doesn't update the .c.qstr file.  And then
> next time round you run make it still tries to generate all qstr
> files again.
I am aware of this behavior and have no good solution to it -
except for allowing the makeqstrdefs.py tool to physically rewrite
the .c.qstr file if asked to do so - when anything on the dependency
path has changed. The problem is actually deeper in the build system
since 'make' does all its decisions based on timestamps. The script
works around this by preventing writing the same content twice.
However, the consequences are the symptoms that we experience as the
dependency path is effectively broken by acting like this - a change in
mpconfigport.h is effectively projected into mpconfig.h that is the
top level dependency for all files. 

A more capable building tool would solve this in a more
robust way using file fingerprints and/or timestamps. I am currently
working on an alternative build system that uses scons, I expect to
publish it in about a week. I am very excited about its performance and
development capabilities. It will allow easy integration of 3rd party
code, configuration management etc. But that's a bit different topic
for now.

Best regards,

Jan

**Comment by dpgeorge** [comment]
> The problem is actually deeper in the build system since 'make' does all its decisions based on timestamps. The script works around this by preventing writing the same content twice. However, the consequences are the symptoms that we experience as the dependency path is effectively broken by acting like this

Yes, that's right, unfortunately timestamps are not sufficient enough for our purposes.

> I am currently working on an alternative build system that uses scons

I don't think we'd move away from makefiles.

To properly fix the problem of auto qstr generation with makefiles we would need to replace the first part of the build with a pure Python script.  It would be forced to run on every make invocation and replicate what make would do for the qstrs: check time stamps, regenerate .c.qstr for those that needed it (and touch those that don't so they don't need checking the next time round), then if anything changed it would rebuild the qstrdefs.h file.  Then make would run as usual.

**Comment by pfalcon** [comment]
@honzik666 : Sorry, but that's over-engineering. https://github.com/micropython/micropython/commit/c618f91e22613b2e530f7c21be0be789a7e8eed5

**Comment by honzik666** [comment]
Dne Tue, 19 Apr 2016 01:43:56 -0700
Paul Sokolovsky <notifications@github.com> napsal(a):

> @honzik666 : Sorry, but that's over-engineering.
> https://github.com/micropython/micropython/commit/c618f91e22613b2e530f7c21be0be789a7e8eed5
Wow, the reworked script parses and relies on a specific preprocessor
output by splitting the .i.last into individual files. 

I would say this solution  does the job and works around the 'make'
deficiency. The readability is maybe a bit lower but that's probably
only my personal preference. I am happy that the feature stays in the
upstream.

Cheers,

Jan

**Comment by honzik666** [comment]
Dne Tue, 19 Apr 2016 01:43:12 -0700
Damien George <notifications@github.com> napsal(a):

> > The problem is actually deeper in the build system since 'make'
> > does all its decisions based on timestamps. The script works around
> > this by preventing writing the same content twice. However, the
> > consequences are the symptoms that we experience as the dependency
> > path is effectively broken by acting like this
> 
> Yes, that's right, unfortunately timestamps are not sufficient enough
> for our purposes.
> 
> > I am currently working on an alternative build system that uses
> > scons
> 
> I don't think we'd move away from makefiles.
ok, good to know.

**Comment by pfalcon** [comment]
> I would say this solution  does the job and works around the 'make'
deficiency.

Alternatively, it just uses it as it was intended ;-). Going away from make, there're so many choices, that it leads to suspicion that there's something wrong with each of them (and knowing some alternatives, like ant, it's easy to believe in that). At least make is simple and its limitations (and how to deal with them) are well known too many people.
